src/server/pfs/server/driver_fsck.go

```diff
@@ -134,163 +134,223 @@ func (e ErrMissingBranchHead) Error() string {
+	// i.e. union(branch, branch.Provenance) = union(branch, branch.DirectProvenance, branch.DirectProvenance.Provenance)
+	direct := bi.DirectProvenance
+	union := []*pfs.Branch{bi.Branch}
+
+	for _, directProvenance := range direct {
+		directProvenanceInfo := branchInfos[pfsdb.BranchKey(directProvenance)]
+		union = append(union, directProvenance)
+		if directProvenanceInfo != nil {
+			union = append(union, directProvenanceInfo.Provenance...)
+		}
+	}
 
-	// TODO(global ids): no fixable fsck issues?
-	// onFix := func(fix string) error { return cb(&pfs.FsckResponse{Fix: fix}) }
+	if !equalBranches(append(bi.Provenance, bi.Branch), union) {
+		if err := onError(ErrBranchProvenanceTransitivity{
+			BranchInfo:     bi,
+			FullProvenance: union,
+		}); err != nil {
+			return err
+		}
+	}
 
-	// collect all the info for the branches and commits in pfs
-	branchInfos := make(map[string]*pfs.BranchInfo)
-	commitInfos := make(map[string]*pfs.CommitInfo)
-	newCommitInfos := make(map[string]*pfs.CommitInfo)
-	repoInfo := &pfs.RepoInfo{}
-	if err := d.repos.ReadOnly(ctx).List(repoInfo, col.DefaultOptions(), func(string) error {
-		commitInfo := &pfs.CommitInfo{}
-		if err := d.commits.ReadOnly(ctx).GetByIndex(pfsdb.CommitsRepoIndex, pfsdb.RepoKey(repoInfo.Repo), commitInfo, col.DefaultOptions(), func(string) error {
-			commitInfos[pfsdb.CommitKey(commitInfo.Commit)] = proto.Clone(commitInfo).(*pfs.CommitInfo)
-			return nil
+	return nil
+}
+
+func checkBranchSubvenances(bi *pfs.BranchInfo, branchInfos map[string]*pfs.BranchInfo, onError func(error) error) error {
+	for _, provBranch := range bi.Provenance {
+		provBranchInfo := branchInfos[pfsdb.BranchKey(provBranch)]
+		if !branchInSet(bi.Branch, provBranchInfo.Subvenance) {
+			if err := onError(ErrBranchSubvenanceTransitivity{
+				BranchInfo:        provBranchInfo,
+				MissingSubvenance: bi.Branch,
+			}); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func checkMissingHead(bi *pfs.BranchInfo, branchInfos map[string]*pfs.BranchInfo, commitInfos map[string]*pfs.CommitInfo, onError func(error) error) error {
+	if bi.Head == nil {
+		if err := onError(ErrMissingBranchHead{
+			Branch: bi.Branch,
 		}); err != nil {
-			return errors.EnsureStack(err)
+			return err
+		}
+	} else {
+		// we expect the branch's provenance to equal the HEAD commit's provenance
+		// i.e branch.Provenance contains the branch provBranch and
+		// provBranch.Head != nil implies branch.Head.Provenance contains
+		// provBranch.Head
+		for _, provBranch := range bi.Provenance {
+			provBranchInfo, ok := branchInfos[pfsdb.BranchKey(provBranch)]
+			if !ok {
+				if err := onError(ErrBranchInfoNotFound{Branch: provBranch}); err != nil {
+					return err
+				}
+				continue
+			}
+			if provBranchInfo.Head != nil {
+				// in this case, the headCommit Provenance should contain provBranch.Head
+				if _, ok := commitInfos[pfsdb.CommitKey(bi.Head)]; !ok {
```

Comment:
isn't the validity of `biHead` enough since`provBranch` is already established in `bi.Head` provenance

Reply:
There is a distinction between branch and commit provenance. I think this check may be outdated since it is going through the full branch provenance. Commits only store their direct provenance. @acohen4 what do you think about this check?

Reply:
This check isn't really doing anything right now. Can we replace it with a check `checkBranchCommitProvenance()` that looks at all of the branch infos, and for each Branch in the `DirectProvenance` look to see that the branch head from the Subvenant branch is Provenant on the branch head commit of the Provenant branch. 

In rough go pseudocode:
`for _, bi := range allBranchInfos {
    assert len(bi.Head.Provenance) == len(bi.DirectProvenance)
    for _, b := range bi.DirectProvenance {
        assert contains(bi.Head.Provenance, getBranchInfo(b).Head)
    }
}`